lib/jara/releaser.rb

```diff
@@ -28,30 +28,34 @@ def initialize(environment, bucket_name, options={})
 
     def build_artifact
       if @environment.nil?
-        jar_name = "#{app_name}.jar"
+        archive_name = "#{app_name}.#{@archiver.extension}"
         Dir.chdir(project_dir) do
-          @archiver.create(jar_name: jar_name)
+          @archiver.create(archive_name: archive_name)
         end
         @logger.info('Created test artifact')
-        File.join(project_dir, 'build', jar_name)
+        File.join(project_dir, 'build', archive_name)
       elsif (artifact_path = find_local_artifact)
         @logger.warn('An artifact for %s already exists: %s' % [branch_sha[0, 8], File.basename(artifact_path)])
         artifact_path
       else
         date_stamp = Time.now.utc.strftime('%Y%m%d%H%M%S')
         destination_dir = File.join(project_dir, 'build', @environment)
-        jar_name = [app_name, @environment, date_stamp, branch_sha[0, 8]].join('-') << '.jar'
+        archive_name = [app_name, @environment, date_stamp, branch_sha[0, 8]].join('-') << '.' << @archiver.extension
         Dir.mktmpdir do |path|
           @shell.exec('git archive --format=tar --prefix=%s/ %s | (cd %s/ && tar xf -)' % [File.basename(path), branch_sha, File.dirname(path)])
           Dir.chdir(path) do
             @logger.info('Checked out %s from branch %s' % [branch_sha[0, 8], @branch])
-            @archiver.create(jar_name: jar_name)
+            if @build_command
+              @logger.info('Running build command: %s' % @build_command)
+              @shell.exec(@build_command)
+            end
```

Comment:
Not really, we bundle gems with extensions, they put their jar files on the class path themselves.

What I mean by general is that the number of things you might want to do is hard to enumerate, but if you get the chance to run a command or some Ruby code on the clean code just before the artifact is built, you can solve many problems, regardless of the type of artifact. Not all, but many, probably most.

Perhaps I'm missing what you mean by making it "a part of the archiver". Do you mean that, for example, Puck is more a build command than an archiver?